auto-value-gson/src/test/java/com/ryanharter/auto/value/gson/AutoValueGsonExtensionTest.java

```diff
@@ -178,35 +171,136 @@ public void simple() {
         + "      }\n"
         + "      jsonWriter.beginObject();\n"
         + "      jsonWriter.name(\"a\");\n"
-        + "      string_adapter.write(jsonWriter, object.a());\n"
+        + "      if (object.a() == null) {\n"
+        + "        jsonWriter.nullValue();\n"
+        + "      } else {\n"
+        + "        if (string_adapter == null) {\n"
+        + "          string_adapter = gson.getAdapter(java.lang.String.class);\n"
+        + "        }\n"
+        + "        string_adapter.write(jsonWriter, object.a());\n"
```

Comment:
I don't like that this creates 3 volatile reads in a row. I would prefer if the pattern
```java
AdapterType string_adapter = this.string_adapter;
if (string_adapter == null) {
  this.string_adapter = string_adapter = gson.getAdapter...
}
string_adapter.write(..)
```
which only requires a single volatile read.

Reply:
Good point. I wonder what the performance effect would be (these operations are unlikely to be contended which I believe increases the cost) but reducing number of volatile reads (I think only, by one, though) will certainly not hurt. Will fix in a separate PR.